### ISSUE TITLE
mbox file import splits emails where a line of body starts with "From"

### ISSUE THREAD

**Issue by pshute** [thundernest/import-export-tools-ng#145]
After importing a large mbox file, I see several emails with messed up senders, all with the date/time of the import.

Investigating a couple of these, it looks like if a line of the body of an email starts with "From", it will interpret it as the start of a new email. It will use the next word on that line as the sender name. I assume because it doesn't find a date header, it inserts the current time.

I was able to reproduce it by editing the mbox file to add the word "From". It has to start with a capital F.

Looks like normally From headers start with "From:", but in this case (export from old version of Forte Agent), they don't have the semi colon. Perhaps this has triggered this behaviour.

**Comment by pshute** [comment]
They all imported correctly after I'd added a space in front of the problem From's.

**Comment by cleidigh** [comment]
@pshute 
The from issue should be handled correctly.
can you tell me the versions?
Christopher

**Comment by pshute** [comment]
Thunderbird 68.9.0 64 bit
ImportExportToolsNG 4.1.0

Here's a sample of a single email that was doing it, obfuscated a bit. Note that the first header has no colon after From, and that there's a second From: header. Not sure if that's normal. In this case the email is imported in two parts, with the split being at "From memory". The second part is imported with the sender as "memory", and the rest of that line lost.

From rad@xxx.org Sat Oct 12 05:36:38 2013
To: LLL <lll@rrr.com.au>
Subject: [xxx] yyyyyyy
From: Russell xxx<rad@xxx.org>
Date: Fri, 11 Oct 2013 20:36:38 +0200

Hi People,

I am trying to track down a friend of mine I have not seen since about 20= 09.

Ron yyy  was formerly a zzz for aaa and then=  moved back to the bbb Area in about 2005.
From memory his wife was an avid qqq

Ron was right into hhh and just in the process of moving up to jjj ga= uge when I last saw him.
If any uuu knows Ron, I'd greatly appreciate you giving him my email a= ddress and the hurry up treatment.

With kind regards and best wishes to all,

Russell xxx
uuu
Austria
--
iii: ppp To unsubscribe send an e-mail to lll@rrr.com.au with the word unsubscribe in the subject: line

**Comment by cleidigh** [comment]
@pshute 
I need to read more slowly, I was thinking of the "From" for handling for EML imports. this would be handled there, however,
we are talking about a mbox files.
I am wondering if there is an assumption that this does not occur within mbox files.
I will check out
Thanks person!
Christopher

**Comment by cleidigh** [comment]
@pshute  
yeah my hunch was correct, Forte uses a slightly different mbox format with an e-mail address after the From separator.
see this post
http://codeverge.com/mozilla.support.thunderbird/import-from-forte-agent/1454848
I am on the fence line dealing with this.  it's the first example of an incompatible mbox file.
a fix would require fully scanning/fixing the separators.
you also mentioned the software was old.
you were savvy enough to detect and even corrects the problem manually most people would not be able to do that
but I wonder how many people have the same issue.
What do you think? do you think this is worth fixing ?
Christopher

**Comment by pshute** [comment]
@cleidigh I don't know what format Agent uses now, or how popular it is. It's likely not many people will be impacted, and not very badly anyway, but I do feel sorry for them.

If you can detect the non standard format, it would be nice if people could be given a warning about what's likely to happen and how to fix it.

**Comment by pshute** [comment]
I'll also mention that before trying your Add-on, I originally tried importing by simply copying the mbox files into the Mail\Local Folders folder, and starting Thunderbird and letting it discover and index them. It did the same thing.

**Comment by cleidigh** [comment]
@pshute 
I will see if I can warning without too much trouble
one thing that would be interesting to know is what the agent does with similar line in the body
Any chance you could create an export and e-mail like that?
Christopher

**Comment by pshute** [comment]
Not sure what you mean. The files I'm trying to import were all exported from Agent.

**Comment by cleidigh** [comment]
I am curious what they do to 'escape' an instance of "From bojo@badboy.com"
within the body
the problem is I can't really do affix without knowing exactly what they do in each circumstance

**Comment by Bravissimo594** [comment]
> I am curious what they do to 'escape' an instance of "From [bojo@badboy.com](mailto:bojo@badboy.com)"
> within the body.

If my memory of running an Email server several years ago serves me correctly, the escape I saw was a greater-than sign (>). This had the effect of making the "From"-started line look like a quote when mail readers not-in-the-know parsed them.

I do not know if it is/was directly part of the SMTP RFC or if it be something MTAs insert, but the envelope sender (MAIL FROM in the SMTP negotiation) showed up as "From foo@bar.com" (without a colon following the "From") in the headers.

You might find more information by following the threads at https://bugzilla.mozilla.org/show_bug.cgi?id=1134690 ... which dealt with (kind of) the opposite problem.

I know this does not answer anything completely, but I hope it helps move the ball forward.

**Comment by pshute** [comment]
I just tried creating an mbox file with gmail, and it added the > prefix for any line starting with From, whether it had an email address in it or not.

I don't want to get this Agent software connected to a mail server again, so I can't create an email to test it. I did see one line in the data I dumped yesterday where the sender had written something like "From the President xyz@abc.com" and that wasn't escaped.

My guess is that either the software would break the same way if I tried to import the files back into it, or they guess from the context. Any line starting with From that doesn't have an email address and a date in it is probably in the body.

Maybe you could present warnings each time such a line is found. A line number would probably help users a lot. I'd be nervous of try a proper fix for fear of messing up standard mbox file imports. Maybe just having it documented here is helpful enough.

**Comment by cleidigh** [comment]
Thunderbird also uses the > when imports or converts messages
I can probably do only message file read to do a probe and possible warning
fixing up the file gets dangerous but probably would have to duplicate safety
have to check this out more once I pass other stuff

**Comment by pshute** [comment]
Thinking more about this, just displaying a warning is probably a good start. If there are more people hitting this problem than expected, the warning might prompt them to register a request for a fix. A lot might not realise they've lost a few emails during imports, or didn't realise it was fixable.

**Comment by cleidigh** [comment]
@pshute 
@Bravissimo594 
The standard practice is to escape beginning line "From" with > as the exporting software
the extension does have a check for formatting, however, it does deal with
several formats of the e-mail separator.  Thunderbird accepts the e-mail type format
as well as just the date format.  So the separator snapped a problem.
Thunderbird/nor the extension deals with escaping the "From memory"
the only way to deal with that is to do a search then import (exactly what  is done for EML)
that will have to wait for a point  release
Christopher

**Comment by ErikKrause** [comment]
I'm having that exact same issue importing mbox files saved from yahoo groups. (Yahoo groups is closing and I need to save the messages from several groups). One line I could track down is 
```
> From 1824 must qualify.
> 
> Sorry personal joke, you have to be Bo to understand 
```
The following lines appear as message body, but without the quote characters. Other occurrences start with a plain 'From' without quote characters. Since those files contain thousands of messages and most false From's contain usual words like 'the', 'my', 'what' etc. it will be hard to fix those in the mbox files.  
I'm using Version 10.0.2 : Maintenance Release - November 14, 2020 currently on TB 68.12.1 (32-Bit)
Please let me know if there is anything I can do to help fixing this.

**Comment by cleidigh** [comment]
@ErikKrause 
thanks for the post.  there seems to be no end to this type of problem!  I will do my best to help.
first questions:
- what is the source of the mbox files?
- are you possibly able to post/e-mail me one or two complete examples of the text within the problematic mbox
?
Christopher

**Comment by ErikKrause** [comment]
Yahoo groups is a mailing list service. They have (or rather had) an export facility which exports all messages from the archive as mbox files in chunks of roughly 10 MB. I can send you such a file privately, but I can not provide it publicly of course. I'll see whether I can prepare an example file later this day...

**Comment by ErikKrause** [comment]
Sorry, I was wrong previously. The lines starting with a quote character (>) don't trigger the error. The relevant text is also present without quotes.
I prepared an example file. I copied three messages which trigger the error and tried to hide the identity of the users by changing personal data: 
[test.zip](https://github.com/thundernest/import-export-tools-ng/files/5568376/test.zip)

**Comment by cleidigh** [comment]
@ErikKrause 
okay great thanks!  working with the files to reproduce the new issue is much easier
it's got to be some small formatting difference for the mbox files. Unfortunately
many of the variations are not standard and require some form of "fixup"
Christopher

**Comment by cleidigh** [comment]
@RebelSoftware 
First sorry about the issue.
I will do my best to help recover your mbox.
There certainly are some issues with Thunderbird and mboxes. I have a totally new import and export mechanism in ietng for Thunderbird v115. The new code does processing to find lines that Thunderbird will misinterpret as separators.
I know you don't want to use it generally, but we may need to use it temporarily to process the file. Then we would bring the mbox back to v102.

Are you interested in some help on this?
@cleidigh

**Comment by RebelSoftware** [comment]
Absolutely I would be interested in some help. 

There is however, no need to apologise, as far as I can tell the issue both you and Thunderbird and other mail clients have is a very poor design for the mbox mail storage format. By not having strict and unique separators, mbox is prone to this sort of issue. The word "From" followed by a space is asking for trouble, I personally cannot see how someone thought it was a good idea.

**Comment by cleidigh** [comment]
@RebelSoftware 
That's great, I really want to make this solid for all your points!
If you don't mind could you tell me your location so I can coordinate communication, I have a shorter window online being disabled.

I think the first thing I would like to investigate is the smallest mbox that fails to import all messages and shows messages with no subject date, etc.

These are usually the From parsing, however, I have specifically added what has been found reliable operating From handling. So if you can find one and hopefully we can analyze it. 
@cleidigh

**Comment by cleidigh** [comment]
In the import that lost mails, were you using Thunderbird 102?
Christopher

**Comment by RebelSoftware** [comment]
I am in South Australia gmt+9:30, I have a few small mbox's that have failed.

I was using 102 when I imported the files. however I upgraded to 115. realised that it had major issues such as missing UI components. It was only after that that I took a copy of my profile and downgraded back to 102. after downgrading was when I made my stupid mistake and inadvertently deleted all emails. 

It was after that that I did the import. so I am not sure if there were changes that 115 made to the mbox that confused 102?
